### PR TITLE
Adding distinct exit codes for cluster connection failures.

### PR DESCRIPTION
**Pull request by PriyaModali**
Fixes: #4645 
**Related**: Tracking issue #4921
**Description**
Adding distinct exit codes for cluster connection failures.

**User facing changes**
These changes will enable Skaffold dev, deploy, to return distinct error codes when kubernetes cluster could not be reached.

Ex. before error messages:
```
Unable to connect to the server: net/http: TLS handshake timeout

exiting dev mode because first deploy failed: unable to connect to Kubernetes: Get "https://192.168.64.3:8443/version?timeout=32s": net/http: TLS handshake timeout
```

Ex. after error messages:
```
Deploy Failed. Could not connect to <cluster name> cluster.

Deploy Failed. Could not connect to cluster test_cluster due to \"https://192.168.64.3:8443/version?timeout=32s\": net/http: TLS handshake timeout. Check your connection for the cluster.
```

### CODE REVIEW

File: pkg/skaffold/errors/deployProblems.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/config"
+	"github.com/GoogleContainerTools/skaffold/proto"
+)
+
+const (
+	// Unable to connect to cluster
+	ClusterConnectErrPrefix = "Could not connect to the cluster."
+)
+
+func suggestDeployClusterConnectonErrorAction(opts config.SkaffoldOptions) []*proto.Suggestion {
```
Comment (tejal29) [inline pkg/skaffold/errors/deployProblems.go:26]:
```suggestion
func suggestDeployClusterConnectonErrorAction(_ config.SkaffoldOptions) []*proto.Suggestion {
```

File: pkg/skaffold/errors/deployProblems.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	if defaultRepo := opts.KubeConfig; defaultRepo != "" {
```
Comment (tejal29) [inline pkg/skaffold/errors/deployProblems.go:27]:
```suggestion

```

File: pkg/skaffold/errors/deployProblems.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+		suggestions := []*proto.Suggestion{{
+			SuggestionCode: proto.SuggestionCode_CHECK_KUBE_CONFIG,
+			Action:         "Check your KubeConfig.",
+		}}
+		return suggestions
+	}
```
Comment (tejal29) [inline pkg/skaffold/errors/deployProblems.go:33]:
```suggestion
```

File: pkg/skaffold/errors/deployProblems.go (hunk 1 above)
Comment (briandealwis) [reply pkg/skaffold/errors/deployProblems.go:26]: Though the code below is using `opts`?

File: pkg/skaffold/errors/err_map.go [hunk 4]
```diff
@@ -87,3 +87,15 @@ var knownBuildProblems = []problem{
 		},
 	},
 }
+
+// Deploy errors in deployment phase
+var knownDeployProblems = []problem{
+	{
+		regexp:  re(fmt.Sprintf(".*%s.* Uanable to connect: .*", ClusterConnectErrPrefix)),
```
Comment (briandealwis) [inline pkg/skaffold/errors/err_map.go:94]:
```suggestion
		regexp:  re(fmt.Sprintf(".*%s.* Unable to connect: .*", ClusterConnectErrPrefix)),
```

File: proto/skaffold.proto [hunk 5]
```diff
@@ -506,6 +506,11 @@ enum StatusCode {
 
     // Deploy cancelled due to user cancellation or one or more deployers failed.
     DEPLOY_CANCELLED = 803;
+
+        // Deploy errors
+
+    // Cluster connection errors
```
Comment (briandealwis) [inline proto/skaffold.proto:512]: Maybe?

```suggestion
    // Unable to connect to cluster
```

File: pkg/skaffold/errors/err_map.go [hunk 6]
```diff
@@ -87,3 +87,15 @@ var knownBuildProblems = []problem{
...
+		errCode: proto.StatusCode_DEPLOY_CLUSTER_CONNECTION_ERR,
+		description: func(error) string {
+			return "Deploy Failed."
```
Comment (briandealwis) [inline pkg/skaffold/errors/err_map.go:97]: I think this gets turned into the log entry, which is then reflected to the output on the IDEs.  Maybe "Unable to connect to cluster"?

File: pkg/skaffold/errors/errors.go [hunk 7]
```diff
@@ -27,28 +27,17 @@ import (
...
 // These are phases in a DevLoop
 const (
-	Init        = Phase("Init")
 	Build       = Phase("Build")
 	Deploy      = Phase("Deploy")
 	StatusCheck = Phase("StatusCheck")
 	FileSync    = Phase("FileSync")
 	DevInit     = Phase("DevInit")
 	Cleanup     = Phase("Cleanup")
-
```
Comment (tejal29) [inline pkg/skaffold/errors/errors.go:37]: you are test failures are due to this unintentional revert

File: pkg/skaffold/errors/errors.go [hunk 8]
```diff
@@ -88,12 +73,24 @@ func ShowAIError(err error) error {
 }
 
 func getErrorCodeFromError(phase Phase, err error) (proto.StatusCode, []*proto.Suggestion) {
-	if problems, ok := allErrors[phase]; ok {
```
Comment (tejal29) [inline pkg/skaffold/errors/errors.go:91]: this is also an intentionalrevert

File: proto/skaffold.proto [hunk 9]
```diff
@@ -560,6 +563,10 @@ enum SuggestionCode {
     // Check if docker is running
     CHECK_DOCKER_RUNNING = 106;
 
+    // Deploy error suggestion codes
+
+    // Kube config error
+    CHECK_KUBE_CONFIG = 201;
```
Comment (tejal29) [inline proto/skaffold.proto:569]: The more description suggestion code would be  `CHECK_CLUSTER_CONNECTION`

**Comment by tejal29** [comment]
@googlebot I consent

File: pkg/skaffold/errors/errors_test.go [hunk 10]
```diff
@@ -243,6 +243,22 @@ func TestShowAIError(t *testing.T) {
 				Suggestions: reportIssueSuggestion(config.SkaffoldOptions{}),
 			},
 		},
+		{
+			description: "deploy failed",
+			opts:        config.SkaffoldOptions{},
+			context:     &config.ContextConfig{},
+			phase:       Deploy,
+			err:         fmt.Errorf(`exiting dev mode because first deploy failed: unable to connect to Kubernetes: Get "https://192.168.64.3:8443/version?timeout=32s": net/http: TLS handshake timeout`),
```
Comment (tejal29) [inline pkg/skaffold/errors/errors_test.go:251]: We should present the cluster ip and also the error details to the user.
e.g.

```
Deploy Failed. Could not connect to cluster gke_tejal-test_us-central1-a_dump  at https://192.168.64.3:8443 due to TLS handshake timeout. Check your cluster connection
```

File: pkg/skaffold/errors/err_map.go (hunk 6 above)
Comment (tejal29) [reply pkg/skaffold/errors/err_map.go:97]: The `description` is not used in the Event API but only for the console. 
The `ActionableErr.Message` which is the raw error is used. 

I made a conscious decision to send the raw error to IDEs to they could do any extra processing of the error message.
Maybe it makes sense to send the `description` or  we could change or add another field to the ` ActionableErr` e.g.
`ProcessedMessage` or `ExtractedMessage` along with the raw error text in `Message`.

File: proto/skaffold.proto (hunk 9 above)
Comment (PriyaModali) [reply proto/skaffold.proto]: Renamed.

File: pkg/skaffold/errors/errors.go (hunk 8 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/errors.go]: Reverted the change.

File: pkg/skaffold/errors/errors.go (hunk 7 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/errors.go]: Reverted the change.

File: proto/skaffold.proto (hunk 5 above)
Comment (PriyaModali) [reply proto/skaffold.proto]: Updated the comment.

File: pkg/skaffold/errors/err_map.go (hunk 4 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/err_map.go]: Fixed the typo.

File: pkg/skaffold/errors/deployProblems.go (hunk 1 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/deployProblems.go]: Fixed.

File: pkg/skaffold/errors/err_map.go (hunk 6 above)
Comment (briandealwis) [reply pkg/skaffold/errors/err_map.go:97]: I only wondered if Skaffold would swallow the `Unable to connect` and instead report `Deploy Failed.`

**Comment by briandealwis** [comment]
BTW: this code path is never run on the first iteration of `dev`.  I tried pausing minikube and then `skaffold dev`.  I see an error reported
```
exiting dev mode because first deploy failed: unable to connect to Kubernetes: Get "https://192.168.64.2:8443/version?timeout=32s": net/http: TLS handshake timeout
```

But there are no events:
```
{"result":{"timestamp":"2020-10-26T17:17:44.630814Z","event":{"metaEvent":{"entry":"Starting Skaffold: \u0026{Version: ConfigVersion:skaffold/v2beta9 GitVersion: GitCommit: GitTreeState: BuildDate: GoVersion:go1.14.3 Compiler:gc Platform:darwin/amd64}","metadata":{"build":{"numberOfArtifacts":1,"builders":[{"type":"DOCKER","count":1}],"type":"LOCAL"},"deploy":{"deployers":[{"type":"KUBECTL","count":1}],"cluster":"MINIKUBE"}}}}}}
{"result":{"timestamp":"2020-10-26T17:17:45.067668Z","event":{"devLoopEvent":{"status":"In Progress"}},"entry":"Update initiated"}}
event.DevLoopFailedInPhase: phase=Deploy deployState=status:"Not Started" autoTrigger:true 
{"result":{"timestamp":"2020-10-26T17:17:55.194247Z","event":{"devLoopEvent":{"status":"Failed","err":{"message":"unable to connect to Kubernetes: Get \"https://192.168.64.2:8443/version?timeout=32s\": net/http: TLS handshake timeout"}}},"entry":"Update failed with error code OK"}}
```

**Comment by tejal29** [comment]
> BTW: this code path is never run on the first iteration of `dev`. I tried pausing minikube and then `skaffold dev`. I see an error reported
> 
> ```
> exiting dev mode because first deploy failed: unable to connect to Kubernetes: Get "https://192.168.64.2:8443/version?timeout=32s": net/http: TLS handshake timeout
> ```
> 
> But there are no events:
> 
> ```
> {"result":{"timestamp":"2020-10-26T17:17:44.630814Z","event":{"metaEvent":{"entry":"Starting Skaffold: \u0026{Version: ConfigVersion:skaffold/v2beta9 GitVersion: GitCommit: GitTreeState: BuildDate: GoVersion:go1.14.3 Compiler:gc Platform:darwin/amd64}","metadata":{"build":{"numberOfArtifacts":1,"builders":[{"type":"DOCKER","count":1}],"type":"LOCAL"},"deploy":{"deployers":[{"type":"KUBECTL","count":1}],"cluster":"MINIKUBE"}}}}}}
> {"result":{"timestamp":"2020-10-26T17:17:45.067668Z","event":{"devLoopEvent":{"status":"In Progress"}},"entry":"Update initiated"}}
> event.DevLoopFailedInPhase: phase=Deploy deployState=status:"Not Started" autoTrigger:true 
> {"result":{"timestamp":"2020-10-26T17:17:55.194247Z","event":{"devLoopEvent":{"status":"Failed","err":{"message":"unable to connect to Kubernetes: Get \"https://192.168.64.2:8443/version?timeout=32s\": net/http: TLS handshake timeout"}}},"entry":"Update failed with error code OK"}}
> ```

@PriyaModali Can you please verify?

File: pkg/skaffold/errors/errors_test.go [hunk 11]
```diff
@@ -243,6 +243,22 @@ func TestShowAIError(t *testing.T) {
...
+			expected:    `exiting dev mode because first deploy failed: unable to connect to Kubernetes: Get "https://192.168.64.3:8443/version?timeout=32s": net/http: TLS handshake timeout`,
```
Comment (tejal29) [inline pkg/skaffold/errors/errors_test.go:252]: Deploy problem should be part of  pkg/skaffold/errors/errors.goL75 and this error message should be 
"Deploy Failed. <More details on why it failed>
Check your cluster connection."

File: pkg/skaffold/errors/err_map.go [hunk 12]
```diff
@@ -87,3 +87,20 @@ var knownBuildProblems = []problem{
 		},
 	},
 }
+
+// Deploy errors in deployment phase
+var knownDeployProblems = []problem{
+	{
+		regexp:  re("(?i).*Unable to connect.*"),
```
Comment (tejal29) [inline pkg/skaffold/errors/err_map.go:94]: you can change this expression to get the error details after Get  in the error message
```suggestion
		regexp:  re("(?i).*unable to connect to Kubernetes: Get (.*)"),
```

File: pkg/skaffold/errors/err_map.go [hunk 13]
```diff
@@ -87,3 +87,20 @@ var knownBuildProblems = []problem{
...
+		errCode: proto.StatusCode_DEPLOY_CLUSTER_CONNECTION_ERR,
+		description: func(error) string {
+			return "Deploy Failed."
```
Comment (tejal29) [inline pkg/skaffold/errors/err_map.go:97]:
```suggestion
description: func(err error) string {
			matchExp := re("(?i).*unable to connect to Kubernetes: Get (.*)")
			if match := matchExp.FindStringSubmatch(fmt.Sprintf("%s", err)); len(match) >= 2 {
				return fmt.Sprintf("Build Failed. Could not connect to cluster %s due to %s", opts.KubeContext, match[1])
			}
			return fmt.Sprintf"Build Failed. Could not connect to Kubernetes cluster.)
		},
```

**Review by tejal29 — CHANGES_REQUESTED** [review]

**Comment by tejal29** [comment]
Currently, we suggest "Check cluster connection" as a suggestion. 

@gsquared94 added a PR to detect if a context is `isMinikube` in this PR #4701. 

We could use this functionality to suggest 
1. If `isMinikube(opts.KubeContext)` is true and `opts.KubeContext == minkube`, then suggestion should be 
```
Check if minikube is running using `minikube status` command and try again
```
2. If `isMinikube(opts.KubeContext)` is true and `opts.KubeContext != minkube` e.g. "cloud-run-dev-internal`, then suggestion should be 
```
Check if minikube is running using `minikube status -p cloud-run-dev-internal` command and try again.
```
3. if `isMinikube(opts.KubeContext)` is false and `opts.KubeContext`== "gke_tejal-test_us-central1-a_dump " then suggestion should be
```
Check your cluster connection for cluster gke_tejal-test_us-central1-a_dump.
```

File: pkg/skaffold/errors/errors_test.go [hunk 14]
```diff
@@ -243,6 +243,22 @@ func TestShowAIError(t *testing.T) {
 				Suggestions: reportIssueSuggestion(config.SkaffoldOptions{}),
 			},
 		},
+		{
+			description: "deploy failed",
+			opts:        config.SkaffoldOptions{},
+			context:     &config.ContextConfig{},
```
Comment (tejal29) [inline pkg/skaffold/errors/errors_test.go:249]: Pass `KubeContext` for different values of context
1. minikube
2. gke.

File: pkg/skaffold/errors/err_map.go [hunk 15]
```diff
@@ -87,3 +87,24 @@ var knownBuildProblems = []problem{
...
+
+// Deploy errors in deployment phase
+var knownDeployProblems = []problem{
+	{
+		regexp:  re("(?i).*unable to connect.*: Get (.*)"),
+		errCode: proto.StatusCode_DEPLOY_CLUSTER_CONNECTION_ERR,
+		description: func(err error) string {
+			matchExp := re("(?i).*unable to connect.*Get (.*)")
+			if match := matchExp.FindStringSubmatch(fmt.Sprintf("%s", err)); len(match) >= 2 {
+				return fmt.Sprintf("Deploy Failed. Could not connect to cluster %s due to %s", skaffoldOpts.KubeContext, match[1])
```
Comment (tejal29) [inline pkg/skaffold/errors/err_map.go:99]: use `GetConfigForCurrentKubectx` to print the active context

File: pkg/skaffold/errors/err_map.go [hunk 16]
```diff
@@ -87,3 +87,24 @@ var knownBuildProblems = []problem{
...
+			}
+			return fmt.Sprintf("Deploy Failed. Could not connect to Kubernetes cluster.")
+		},
+		suggestion: func(opts config.SkaffoldOptions) []*proto.Suggestion {
+			return []*proto.Suggestion{{
```
Comment (tejal29) [inline pkg/skaffold/errors/err_map.go:104]: use `util.IsMinikube` to suggest different actions like i mentioned in https://github.com/GoogleContainerTools/skaffold/pull/4933#issuecomment-716888249

**Review by tejal29 — CHANGES_REQUESTED** [review]

File: pkg/skaffold/errors/errors_test.go (hunk 10 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/errors_test.go:251]: Fixed.

File: pkg/skaffold/errors/errors_test.go (hunk 11 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/errors_test.go]: Fixed.

File: pkg/skaffold/errors/err_map.go (hunk 13 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/err_map.go]: Addressed.

File: pkg/skaffold/errors/errors_test.go (hunk 14 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/errors_test.go:249]: Updated.

File: pkg/skaffold/errors/err_map.go (hunk 15 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/err_map.go]: Updated code to retrieve the active context with kubectx.

File: pkg/skaffold/errors/err_map.go [hunk 17]
```diff
@@ -87,3 +88,26 @@ var knownBuildProblems = []problem{
...
 }
+
+// Deploy errors in deployment phase
+var knownDeployProblems = []problem{
+	{
+		regexp:  re("(?i).*unable to connect.*: Get (.*)"),
+		errCode: proto.StatusCode_DEPLOY_CLUSTER_CONNECTION_ERR,
+		description: func(err error) string {
+			matchExp := re("(?i).*unable to connect.*Get (.*)")
+			kubeconfig, _ := kubectx.CurrentConfig()
```
Comment (tejal29) [inline pkg/skaffold/errors/err_map.go:99]: don't ignore the error. Please log it at debug level using `logrus.Debugf`

File: pkg/skaffold/errors/err_map.go [hunk 18]
```diff
@@ -87,3 +88,26 @@ var knownBuildProblems = []problem{
...
+			if match := matchExp.FindStringSubmatch(fmt.Sprintf("%s", err)); len(match) >= 2 {
+				// kubeconfig, _ := kubectx.CurrentConfig()
+				return fmt.Sprintf("Deploy Failed. Could not connect to cluster %s due to %s", kubeconfig.CurrentContext, match[1])
+			}
+			return fmt.Sprintf("Deploy Failed. Could not connect to %s cluster.", kubeconfig.CurrentContext)
+		},
+		suggestion: func(opts config.SkaffoldOptions) []*proto.Suggestion {
+			return []*proto.Suggestion{{
+				SuggestionCode: proto.SuggestionCode_CHECK_HOST_CONNECTION,
```
Comment (tejal29) [inline pkg/skaffold/errors/err_map.go:108]: Can we use the same function `kubectx.CurrentConfig()` to determine the active context and suggest different actions when context is minikube vs GKE.
you can use the `cluster.isMinikube` method to determine if this is a minikube cluster

File: proto/skaffold.proto [hunk 19]
```diff
@@ -560,6 +562,10 @@ enum SuggestionCode {
     // Check if docker is running
     CHECK_DOCKER_RUNNING = 106;
 
+    // Deploy error suggestion codes
+
+    // Check cluster connection
+    CHECK_CLUSTER_CONNECTION = 201;
```
Comment (tejal29) [inline proto/skaffold.proto:568]: Please add a new suggestion code  `CHECK_MINIKUBE_STAUTUS` and use that if cluster is minikube

File: pkg/skaffold/errors/errors_test.go [hunk 20]
```diff
@@ -243,6 +243,22 @@ func TestShowAIError(t *testing.T) {
 				Suggestions: reportIssueSuggestion(config.SkaffoldOptions{}),
 			},
 		},
+		{
+			description: "deploy failed",
+			opts:        config.SkaffoldOptions{},
```
Comment (tejal29) [inline pkg/skaffold/errors/errors_test.go:248]: you need to pass in different contexts in `context` field, 
1. Minikube
2. Non Local cluster or GKE

The config.ContextConfig object has a `Kubecontext` 
1. Setting the value to "minikube" will test  local minikube scenario
2. Setting the value to something other than "minikube" will test the remote cluster scenario.

File: pkg/skaffold/errors/errors_test.go (hunk 14 above)
Comment (tejal29) [reply pkg/skaffold/errors/errors_test.go:249]: you have to pass the value to the unit tests.

**Review by tejal29 — CHANGES_REQUESTED** [review]

File: pkg/skaffold/errors/deploy_problems.go [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+	// for testing
+	currentConfig = kubectx.CurrentConfig
+)
+
+func suggestDeployFailedAction(opts config.SkaffoldOptions) []*proto.Suggestion {
+	kubeconfig, parsederr := currentConfig()
+	logrus.Debugf("Error retrieving the config: %q", parsederr)
+
+	var curctx = kubeconfig.CurrentContext
+	var isminikube = cluster.GetClient().IsMinikube(opts.KubeContext)
```
Comment (tejal29) [inline pkg/skaffold/errors/deploy_problems.go:37]: the `curCtx` evaluates the current active deploy context i.e. is the cluster you are deploying to. 
The code needs to check if the evaluated `isMinikube` and not value specified in `opts.KubeContext`

File: pkg/skaffold/errors/deploy_problems.go [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+	var kubectx = opts.KubeContext
```
Comment (tejal29) [inline pkg/skaffold/errors/deploy_problems.go:38]: this is not needed.

File: pkg/skaffold/errors/err_map.go [hunk 23]
```diff
@@ -22,8 +22,14 @@ import (
 
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/config"
 	"github.com/GoogleContainerTools/skaffold/proto"
+	"github.com/sirupsen/logrus"
 )
 
+// var (
```
Comment (tejal29) [inline pkg/skaffold/errors/err_map.go:28]: please remove uncommented code.

File: pkg/skaffold/errors/err_map.go [hunk 24]
```diff
@@ -87,3 +93,21 @@ var knownBuildProblems = []problem{
...
+
+// Deploy errors in deployment phase
+var knownDeployProblems = []problem{
+	{
+		regexp:  re("(?i).*unable to connect.*: Get (.*)"),
+		errCode: proto.StatusCode_DEPLOY_CLUSTER_CONNECTION_ERR,
+		description: func(err error) string {
+			matchExp := re("(?i).*unable to connect.*Get (.*)")
+			kubeconfig, parsederr := currentConfig()
+			logrus.Debugf("Error retrieving the config: %q", parsederr)
```
Comment (tejal29) [inline pkg/skaffold/errors/err_map.go:105]: only log if err != nil

File: pkg/skaffold/errors/deploy_problems.go (hunk 21 above)
Comment (tejal29) [inline pkg/skaffold/errors/deploy_problems.go:37]: I am surprised if the unit test is working. You need to mock this function call in your tests.

**Review by tejal29 — CHANGES_REQUESTED** [review]
1. Use evaluated kubeContext to check if cluster is local/minikube.
2. Mock `isMinkube` and add all variations of tests. 
- isMinikube=true, clusterName = minikube
-  isMinikube=true, clusterName = some_random
-  isMinikube=false, cluterName = 'test_cluster"  (what you have right now)

**Comment by tejal29** [comment]
> Currently, we suggest "Check cluster connection" as a suggestion.
> 
> @gsquared94 added a PR to detect if a context is `isMinikube` in this PR #4701.
> 
> We could use this functionality to suggest
> 
> 1. If `cluster.IsMinikube(opts.KubeContext)` is true and `opts.KubeContext == minkube`, then suggestion should be
> 
> ```
> Check if minikube is running using `minikube status` command and try again
> ```
> 
> 1. If `cluster.IsMinikube(opts.KubeContext)` is true and `opts.KubeContext != minkube` e.g. "cloud-run-dev-internal`, then suggestion should be
> 
> ```
> Check if minikube is running using `minikube status -p cloud-run-dev-internal` command and try again.
> ```
> 
> 1. if `cluster.IsMinikube(opts.KubeContext)` is false and `opts.KubeContext`== "gke_tejal-test_us-central1-a_dump " then suggestion should be
> 
> ```
> Check your cluster connection for cluster gke_tejal-test_us-central1-a_dump.
> ```
> 
> It would be great, if for conditions 1 and 2, if we can actually run `minikube status` command and
> 
> 1. if minikube is paused, suggest users to run `minikube unpause -p <>`
> 2. if minikube stopped, then suggest users to run `minikube start -p <>`.
> 
> You can follow up the above two changes in another PR and create an issue for this.

We spoke in our 1:1 on Friday, to switch `opts.KubeContext` with the `currentConfig()` value

File: pkg/skaffold/errors/deploy_problems.go (hunk 22 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/deploy_problems.go:38]: Removed.

File: pkg/skaffold/errors/err_map.go (hunk 24 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/err_map.go:105]: Updated.

File: pkg/skaffold/errors/err_map.go (hunk 18 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/err_map.go]: Updated.

File: pkg/skaffold/errors/err_map.go (hunk 24 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/err_map.go:105]: done.

File: pkg/skaffold/errors/deploy_problems.go (hunk 21 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/deploy_problems.go:37]: The unit test is working. :-)

File: pkg/skaffold/errors/err_map.go (hunk 23 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/err_map.go:28]: Removed.

File: pkg/skaffold/errors/deploy_problems.go (hunk 21 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/deploy_problems.go:37]: Updated.

File: pkg/skaffold/errors/deploy_problems.go [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+			Action:         "Check if minikube is running using `minikube status -p <>` command and try again.",
+		}}
+	}
+	if isminikube && kubectx != "minkube" {
+		// Check your cluster connection for your named cluster.
+		return []*proto.Suggestion{{
+			SuggestionCode: proto.SuggestionCode_CHECK_CLUSTER_CONNECTION,
+			Action:         "Check if minikube is running using `minikube status` command and try again",
+		}}
+	}
```
Comment (tejal29) [inline pkg/skaffold/errors/deploy_problems.go:60]:
```suggestion
	if isminikube:= cluster.GetClient().IsMinikube(kubeconfig.CurrentContext()) && kubeconfig.CurrentContext() == minikube {
		return []*proto.Suggestion{{
			SuggestionCode: proto.SuggestionCode_CHECK_MINIKUBE_STAUTUS,
			Action:         "Check if minikube is running using `minikube status` command and try again",
		}}
	}else if isminikube  {
		return []*proto.Suggestion{{
			SuggestionCode: proto.SuggestionCode_CHECK_MINIKUBE_STAUTUS,
			Action:         "Check if minikube is running using `minikube status -p %s` command and try again.",
		}}
	} else {
		return []*proto.Suggestion{{
			SuggestionCode: proto.SuggestionCode_CHECK_CLUSTER_CONNECTION,
			Action:         "Check if minikube is running using `minikube status` command and try again",
		}}
	}
```

File: pkg/skaffold/errors/deploy_problems.go [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+
+	return []*proto.Suggestion{{
+		SuggestionCode: proto.SuggestionCode_CHECK_CLUSTER_CONNECTION,
+		Action:         "Check your cluster connection",
+	}}
```
Comment (tejal29) [inline pkg/skaffold/errors/deploy_problems.go:65]: you dont need this anymore
```suggestion
```

**Review by tejal29 — CHANGES_REQUESTED** [review]

File: pkg/skaffold/errors/deploy_problems.go (hunk 21 above)
Comment (tejal29) [reply pkg/skaffold/errors/deploy_problems.go:37]: hmm. That is because you are not testing other variations i.e. cluster is a local cluster in you test.

**Comment by tejal29** [comment]
this is top level comment.

File: pkg/skaffold/errors/deploy_problems.go [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+	kubeconfig, parsederr := kubectx.CurrentConfig()
+	if parsederr != nil {
+		logrus.Debugf("Error retrieving the config: %q", parsederr)
+		return []*proto.Suggestion{{
+			SuggestionCode: proto.SuggestionCode_CHECK_CLUSTER_CONNECTION,
+			Action:         "Check your connection for the cluster",
+		}}
+	}
+
+	var curctx = kubeconfig.CurrentContext
```
Comment (tejal29) [inline pkg/skaffold/errors/deploy_problems.go:44]: drop this variable and use `kubeconfig.CurrentContext` as is

File: pkg/skaffold/errors/deploy_problems.go [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+	var isminikube = isMinikube(curctx)
+
+	if isminikube {
+		if curctx == "minikube" {
```
Comment (tejal29) [inline pkg/skaffold/errors/deploy_problems.go:48]: you can replace this with a inline variable and one if statement. 
Please see comment https://github.com/GoogleContainerTools/skaffold/pull/4933#discussion_r516281783

**Review by tejal29 — COMMENTED** [review]
This code Looks good to me expect for 1 nit.

File: pkg/skaffold/errors/deploy_problems.go (hunk 27 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/deploy_problems.go:44]: `curctx` is being used at multiple places (3 different places)

**Review by tejal29 — APPROVED** [review]

File: pkg/skaffold/errors/err_map.go (hunk 6 above)
Comment (tejal29) [reply pkg/skaffold/errors/err_map.go]: ah. got it. something like this
```
Deploy Failed. Check your connection for the cluster test_cluster.
```
Vs what we have right now. 

```

Deploy Failed. Could not connect to cluster test_cluster due to \"https://192.168.64.3:8443/version?timeout=32s\": net/http: TLS handshake timeout. 

Check your connection for the cluster.
```

File: pkg/skaffold/errors/deploy_problems.go (hunk 27 above)
Comment (tejal29) [reply pkg/skaffold/errors/deploy_problems.go]: in this case, `kubeconfig.CurrentContext` is a struct field. its a evaluated function and hence it is ok to re-use it.

File: pkg/skaffold/errors/err_map.go (hunk 6 above)
Comment (PriyaModali) [reply pkg/skaffold/errors/err_map.go]: This change makes the message more readable / actionable.

**Pull request merged**